src/stdlib_stringlist.f90

```diff
@@ -0,0 +1,726 @@
+        type(stringlist_type), intent(in)           :: list
+
+        eq_carray_stringlist = ( list == carray )
+
+    end function eq_carray_stringlist
+
+    !> Compares stringlist 'list' for equality with stringarray 'sarray'
+    !> Returns a logical
+    pure logical function eq_sarray_stringlist( sarray, list )
+        type(string_type), dimension(:), intent(in) :: sarray
+        type(stringlist_type), intent(in)           :: list
+
+        eq_sarray_stringlist = ( list == sarray )
+
+    end function eq_sarray_stringlist
+
+    !> Compares stringlist 'list' for inequality with stringlist 'slist'
+    !> Returns a logical
+    pure logical function ineq_stringlist( list, slist )
+        type(stringlist_type), intent(in)           :: list
+        type(stringlist_type), intent(in)           :: slist
+
+        ineq_stringlist = .not.( list == slist )
+
+    end function ineq_stringlist
+
+    !> Compares stringlist 'list' for inequality with chararray 'carray'
+    !> Returns a logical
+    pure logical function ineq_stringlist_carray( list, carray )
+        type(stringlist_type), intent(in)           :: list
+        character(len=*), dimension(:), intent(in)  :: carray
+
+        ineq_stringlist_carray = .not.( list == carray ) 
+
+    end function ineq_stringlist_carray
+
+    !> Compares stringlist 'list' for inequality with stringarray 'sarray'
+    !> Returns a logical
+    pure logical function ineq_stringlist_sarray( list, sarray )
+        type(stringlist_type), intent(in)           :: list
+        type(string_type), dimension(:), intent(in) :: sarray
+
+        ineq_stringlist_sarray = .not.( list == sarray ) 
+
+    end function ineq_stringlist_sarray
+
+    !> Compares stringlist 'list' for inequality with chararray 'carray'
+    !> Returns a logical
+    pure logical function ineq_carray_stringlist( carray, list )
+        character(len=*), dimension(:), intent(in)  :: carray
+        type(stringlist_type), intent(in)           :: list
+
+        ineq_carray_stringlist = .not.( carray == list)
+
+    end function ineq_carray_stringlist
+
+    !> Compares stringlist 'list' for inequality with stringarray 'sarray'
+    !> Returns a logical
+    pure logical function ineq_sarray_stringlist( sarray, list )
+        type(string_type), dimension(:), intent(in) :: sarray
+        type(stringlist_type), intent(in)           :: list
+
+        ineq_sarray_stringlist = .not.( sarray == list )
+
+    end function ineq_sarray_stringlist
+
+  ! destroy:
+
+    !> Version: experimental
+    !>
+    !> Resets stringlist 'list' to an empy stringlist of len 0
+    !> Modifies the input stringlist 'list'
+    subroutine destroy_list( list )
+        !> TODO: needs a better name?? like clear_list or reset_list
+        class(stringlist_type), intent(out) :: list
+
+        list%size = 0
+        if ( allocated( list%stringarray ) ) then
+            deallocate( list%stringarray )
+        end if
+
+    end subroutine destroy_list
```

Comment:
The `intent(out)` here means that `list` will always have `list%size = 0` and `.not. allocated(list%stringarray)` at the start of the procedure.

Reply:
I thought `intent(out)` mean that I won't read anything from the input argument `list` and only write in it.

Reply:
`intent(out)` does mean that, but it also means that the actual argument (e.g. on the calling side) is finalized (either an explicit finalizer is called on the object, or it is returned to a state "as if" no operations had ever been performed on it) before execution of the procedure begins.

Reply:
From the standard:

> The INTENT (OUT) attribute for a nonpointer dummy argument specifies that the dummy argument becomes
undefined on invocation of the procedure, except for any subcomponents that are default-initialized (7.5.4.6). Any
actual argument that corresponds to such a dummy argument shall be definable. The INTENT (OUT) attribute
for a pointer dummy argument specifies that on invocation of the procedure the pointer association status of
the dummy argument becomes undefined. Any actual argument that corresponds to such a pointer dummy shall
be a pointer variable or a procedure pointer that is not the result of a function reference. Any undefinition or
definition implied by association of an actual argument with an INTENT (OUT) dummy argument shall not
affect any other entity within the statement that invokes the procedure.

Reply:
So, how should I solve this problem now:
1). remove line 417:419
2). change `intent(out)` to `intent(inout)`

Reply:
Using `intent(inout)` seems preferable.

Reply:
I just realised for **gfortran** that when I use `intent(inout)` it deallocated `stringarray`.

Reply:
Isn't that the expected behaviour here and the same you are getting with `intent(out)`. Destroying the string list should reset it, which means that the list of stored strings is destroyed as well. The string list will have a capacity of zero afterwards and will get reallocated at the next insertion.

Reply:
I meant that when I removed line 417:419 and used `intent(inout)` (instead of `intent(out)`). After doing this I expected failures in test cases but they all passed because I thought `list` will not get reset but it did for **gfortran**.

Reply:
Sounds like a stale build artifact somewhere, try to run the clean target and compile again.

Reply:
You still set size to zero, but keep the buffer allocated. This would keep the stale buffer around unless you reallocate it at an insertion. Since size is zero you can't access the stale buffer anymore.

Still, it is better to drop the buffer as well. Destroying the list shouldn't only loose the references to the entries but also free the memory they occupy.

Reply:
I would just do 1. Leave it as `intent(out)`, and the compiler does all the work for you.

Reply:
At least Intel will be generating warnings for the `intent(out)` only case. Alternatively we could recommend to destroy a list to recreate it with:

```f90
list = stringlist_type()
```

Reply:
Looking at the code above, I don't think that any such finalizer is necessary for allocatable subcomponents. In any case, such a routine (from the intention) needs to have intent(inout).

Reply:
It's not meant as `final` procedure, but to provide a way to clear all entries from the list beside the default finalizer when leaving a scope. Maybe the name `destroy` is not the best choice here.

Reply:
> such a routine (from the intention) needs to have intent(inout).

That makes sense. I agree.

Looking at the TODO there, I think `clear` is the most applicable name, and matches well with the `intent(inout)`.